server/sonar-db-migration/src/main/java/org/sonar/server/platform/db/migration/version/v71/MigrateWebhooksToWebhooksTable.java

```diff
@@ -0,0 +1,316 @@
+      deleteFromProperties(context, toBeDeleted);
+    }
+
+  }
+
+  private static Map<Long, String> getProjectsUuids(Context context) throws SQLException {
+    return context
+      .prepareSelect("select id, uuid from projects")
+      .list(row -> new ProjectFlyweight(
+        row.getLong(1),
+        row.getString(2)))
+      .stream()
+      .collect(toMap(ProjectFlyweight::id, ProjectFlyweight::uuid));
+  }
+
+  private List<Long> migrateProjectsWebhooks(Context context, Map<String, PropertyRow> properties, Map<Long, String> projectsKeys) throws SQLException {
+    List<Long> toBeDeleted = new ArrayList<>();
+    PropertyRow index = properties.get("sonar.webhooks.project");
+    if (index == null) {
+      return emptyList();
+    }
+
+    // can't lambda due to checked exception.
+    for (Webhook webhook : extractProjectWebhooksFrom(properties, index.value().split(","), projectsKeys)) {
+      toBeDeleted.addAll(insert(context, webhook));
+    }
+
+    toBeDeleted.add(index.id());
+    return toBeDeleted;
+  }
+
+  private List<Long> migrateGlobalWebhooks(Context context, Map<String, PropertyRow> properties) throws SQLException {
+
+    List<Long> toBeDeleted = new ArrayList<>();
+    PropertyRow index = properties.get("sonar.webhooks.global");
+    if (index == null) {
+      return emptyList();
+    }
+
+    // can't lambda due to checked exception.
+    for (Webhook webhook : extractGlobalWebhooksFrom(context, properties, index.value().split(","))) {
+      toBeDeleted.addAll(insert(context, webhook));
+    }
+
+    toBeDeleted.add(index.id());
+    return toBeDeleted;
+  }
+
+  private List<Long> insert(Context context, Webhook webhook) throws SQLException {
+
+    if (webhook.isValid()) {
+      context.prepareUpsert("insert into webhooks (uuid, name, url, organization_uuid, project_uuid, created_at, updated_at) values (?, ?, ?, ?, ?, ?, ?)")
+        .setString(1, uuidFactory.create())
+        .setString(2, webhook.name())
+        .setString(3, webhook.url())
+        .setString(4, webhook.organisationUuid())
+        .setString(5, webhook.projectUuid())
+        .setLong(6, webhook.createdAt())
+        .setLong(7, webhook.createdAt())
+        .execute()
+        .commit();
+    } else {
+      LOGGER.info("Unable to migrate inconsistent webhook (entry deleted from PROPERTIES) : " + webhook);
+    }
+
+    return webhook.rowIds();
+  }
+
+  private static void deleteFromProperties(Context context, List<Long> toBeDeleted) throws SQLException {
+    Upsert upsert = context.prepareUpsert(prepareDeleteRequest(toBeDeleted));
+    int index = 0;
+    for (Long id : toBeDeleted) {
+      index++;
+      upsert.setLong(index, id);
+    }
+    upsert.execute().commit();
+  }
+
+  private static String prepareDeleteRequest(List<Long> toBeDeleted) {
+    StringBuilder sb = new StringBuilder()
+      .append("delete from properties where id in (");
```

Comment:
That fails on Oracle if there are more than 1'000 ids (and on MSSQL if more than 2'000). A solution is to replace by `delete from properties where prop_key like 'sonar.webhooks.global%' or prop_key like 'sonar.webhooks.project%'`

Reply:
much more efficient + it removes a bunch of code (collecting ids to delete) done